sdk/appconfiguration/azure-spring-cloud-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/refresh/CalculatedBackoffTime.java

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.spring.cloud.config.refresh;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.concurrent.TimeUnit;
+
+import com.azure.spring.cloud.config.properties.AppConfigurationProviderProperties;
+
+/**
+ * Calculates the amount of time to the next refresh, if a refresh fails.
+ */
+public final class CalculatedBackoffTime {
+
+    static HashMap<String, Integer> attempts = new HashMap<>();
+
+    /**
+     * Sets the number of retry attempts back to 0.
+     */
+    public static void resetAttempts() {
+        attempts = new HashMap<>();
+    }
+
+    /**
+     * Calculates the amount of time to the next refresh, if a refresh fails. Takes current Refresh date into account
+     * for watch keys.
+     * @param endpoint value for tracking the failed attempt number
+     * @param currentRefresh the current refresh date
+     * @param interval the Refresh Interval
+     * @param properties App Configuration Provider Properties
+     * @return new Refresh Date
+     */
+    public static Instant calculateBefore(String endpoint, Instant currentRefresh, Duration interval,
+        AppConfigurationProviderProperties properties) {
+
+        if (interval == null) {
+            return null;
+        }
+
+        // The refresh interval is only updated if it is expired.
+        if (Instant.now().isAfter(currentRefresh)) {
+            Integer attempt = attempts.getOrDefault(endpoint, 1);
+            attempts.put(endpoint, attempt + 1);
+
+            int durationPeriod = Math.toIntExact(interval.getSeconds());
+
+            Instant now = Instant.now();
+
+            if (durationPeriod <= properties.getDefaultMinBackoff()) {
+                return now.plusSeconds(interval.getSeconds());
+            }
+
+            long defaultMinBackoff = TimeUnit.NANOSECONDS.convert(properties.getDefaultMinBackoff(), TimeUnit.SECONDS);
+            double maxRandomValue = Math.min(Math.pow(2, attempt - 1), Double.MAX_VALUE);
```

Comment:
We should avoid using `Math.pow` since that can overflow. Instead, you can use left shift operator to calculate power of 2. This is the C# equivalent, where MaxAttempts is 63:
`(long)1 << Math.Min(attempts, MaxAttempts)`

Reply:
`Math.pow` doesn't overflow, just switches to scientific notation and loses precision. I don't think it will be an issue.

Reply:
Bitwise operation is still faster than Math.Pow. There is scope for performance improvement here since this backoff calculation will be done whenever refresh operation fails.